### PR TITLE
nn.Module hook fix and improvements

### PR DESCRIPTION
**Pull request by albanD**
This PR adds the following:
- Allow forward pre hooks and foward hooks to modify the input by returning a not-None value. Some checks are done here to make sure the returned value is similar to the input.
- Fix the backward hooks for the case where all inputs and outputs are Tensors. Raise an error if any input or output is not a Tensor.

 Do we actually want to check the returned values from the forward hooks?

Adding support for Module taking other things than Tensors (and giving `None` gradient for these) is left to be done later.
This is a breaking change with the current nn.Module backward hooks. But no one should be using them as they are heavily broken atm anyway.

### CODE REVIEW

**Comment by ezyang** [comment]
Hey @albanD, could you rebase this on master? That will get CircleCI tests working for you.

**Comment by ezyang** [comment]
Sorry, this might take a little time to review. @apaszke, do you think you have time to look at this? Otherwise I'll try to squeeze some in later.

File: torch/utils/hooks.py [hunk 1]
```diff
@@ -34,3 +34,22 @@ def __enter__(self):
...
+    def __init__(self, module, user_hook):
+        self.grad_outputs = None
+        self.user_hook = user_hook
+        self.module = module
+
+    def get_input_hook(self):
+        def hook(grad_input, _):
+            return self.user_hook(self.module, grad_input, self.grad_output)
+        # Make error message more user-friendly
+        hook.__name__ = self.user_hook.__name__
```
Comment (SsnL) [inline torch/utils/hooks.py]: use `@functools.wraps` :)

File: torch/autograd/_functions/tensor.py [hunk 2]
```diff
@@ -21,6 +21,16 @@ def backward(ctx, grad_output):
                 return grad_output.type(ctx.input_type), None
 
 
+class Noop(Function):
```
Comment (SsnL) [inline torch/autograd/_functions/tensor.py]: @colesbury How does Python autograd function work with our View logic? My guess is that this function is similar to an `input[:]` except that the output only shares storage but not has ViewImpl (a non-differentiable view if you use the terminology in #12502 ). I'm mainly worried that `rebase_history` of base will move this Function around so that the hooks registers on this is not the accurate grad_input/output. I'll need to think about this more.

**Comment by albanD** [comment]
@ezyang CircleCI tests were failing because weird git related issues when it was pulling the branch. So I rebased on top of master. Not sure what's happening :/

**Comment by yf225** [comment]
@albanD there are some CircleCI changes since your rebase, and we might need to rebase again or pull master into your branch to have it working

**Comment by yf225** [comment]
@ezyang it seems that CircleCI has history of it (https://circleci.com/gh/pytorch/workflows/pytorch/tree/pull%2F12573), but the statuses don't show up here. I just submitted a ticket to CircleCI.

**Comment by albanD** [comment]
@yf225 I did the rebase, seems to work now !

**Comment by apaszke** [comment]
@ezyang yes, I'd like to take a look at this to make sure everything is sound. I should have some time tomorrow in the evening.

**Review by apaszke — CHANGES_REQUESTED** [review]
I have some suggestions for improvements, but the PR generally looks ok. It's definitely a step in a good direction compared to the fully broken behavior we have today, but I'm worried that it might still fail in cases like this one (that's related to resetting of `grad_fn`s of views):
```
input = torch.randn(2, 5, requires_grad=True)
module = nn.Linear(5, 10)
module.register_backward_hook(lambda *args: print('Hook!')
output = module(input)
# Now do either
# input.mul_(2)
# output.mul_(2)
output.sum().backward()
```

File: torch/nn/modules/module.py [hunk 3]
```diff
@@ -468,32 +512,59 @@ def _slow_forward(self, *input, **kwargs):
...
+            noop_fn = _get_prev_function(input)
+            if noop_fn is not None:
+                for hook in backward_hooks:
+                    noop_fn.register_hook(hook.get_input_hook())
+
         for hook in self._forward_pre_hooks.values():
-            hook(self, input)
+            hook_result = hook(self, input)
+            if hook_result is not None:
+                _check_same_shape(input, hook_result, "forward pre hook '{}'".format(hook))
```
[the remark this answers is not in the document]
Comment (albanD) [reply torch/nn/modules/module.py]: I'm happy to drop these.

File: torch/nn/modules/module.py [hunk 4]
```diff
@@ -20,6 +20,43 @@ def _addindent(s_, numSpaces):
     return s
 
 
+def _check_same_shape(base, new, caller_name):
```
[the remark this answers is not in the document]
Comment (albanD) [reply torch/nn/modules/module.py]: This is less strict than that actually.
It only checks that each element for `*args` has the same type and for tensors the same size.
Maybe the function name is misleading. The goal was to make sure that arguments where not completely changed.
But we can drop this if you prefer?

**Comment by albanD** [comment]
This strategy is promising but cause a lot of problems if inputs or outputs are modified inplace. In particular, hooks are not called anymore.
The behavior of hooks will need to be cleaned up before we can make this possible, closing this PR in the meantime.

**Comment by NarineK** [comment]
@albanD, any news on improving the hooks ?

**Comment by albanD** [comment]
Hi,
Unfortunately no, nothing new at the moment as some underlying changes are required in the autograd engine before this strategy can be used for the backward hooks.
If you're looking for forward hooks, they should work fine. If you want to be able able to change the output of forward hooks, that change can be done quite easily. Let me know and i'll extract it from this PR and make a new one just with this. Or maybe you want something else with the forward hooks?

**Comment by NarineK** [comment]
Hi @albanD! Sorry for the late reply. I somehow missed the notification and I've just seen your comment. 
I'm currently more concerned about the backward hooks of PyTorch modules and the inconsistencies mentioned here: https://github.com/pytorch/pytorch/issues/12331
Is there any documentation about the inconsistencies that one can expect for the `grad_input` and `grad_output`. In some of the issues @apaszke  mentioned that you might want to remove the support for backward hooks. Perhaps for the container type modules only ?
